### ISSUE TITLE
Pipeline GUI Automated Testing

### ISSUE THREAD

**Issue by tomspilman** [mono/MonoGame#4558]
We need to develop or find a solution to effectively automate the testing of the GUI Pipeline tool.  We should be able to write scripts/record macros that perform common tasks, have the test runner execute them, and verify the GUI produced the expected results.

This is critical in making the tool reliable.

Ideally the solution is cross platform and we can run most of the same tests without modifications on Mac and Linux as well.

**Comment by tomspilman** [comment]
If we were to roll our own system we would need to add a new interface abstraction into the pipeline tool that can do stuff like:

 - Change the selection/focus based on control ID or relative position.
 - Send keyboard/mouse events to the selected control/window.
 - Be able to read the value from text and combo boxes.
 - Get events for dialogs appearing/disappearing.
 - Open/close/trigger menus.
 - Be able to read a window title.

Once we have a system like this we can code a unit test in that drives the UI.  Best of all the tests should work cross platform and ensure we never break functionality in the Pipeline tool.

**Comment by tomspilman** [comment]
Microsoft has a system for this called "Coded UI Tests", but it is windows specific... so it can't solve our case.  Still it is useful to see how it works and what the code it generates looks like:  https://msdn.microsoft.com/en-us/library/ff977233.aspx

**Comment by cra0zy** [comment]
I know how we can do partial testing, we can call the Controller and execute some basic actions, it won't really tell us if GUI looks alright, but most errors occur because of the Controller anyway so it will be really helpful. Thoughts?

**Comment by tomspilman** [comment]
> we can call the Controller and execute some basic actions

Yeah...  certainly it is better than the complete lack of testing we have now.

We would probably want this as some separate test project specifically for the pipeline tool.

**Comment by tomspilman** [comment]
@cra0zy 

We need to start work on this as soon as possible.  Specifically what I outlined above in https://github.com/MonoGame/MonoGame/issues/4558#issuecomment-208164178.

We will not add any new features to the pipeline tool until we get some automated testing in place.

**Comment by cra0zy** [comment]
To continue the discussion from #5694:

> We can't have a test solution that only works on Windows.
>
> The only choice I think we have is to roll our own solution that does specifically what we need.

I know how to setup GUI testing for Windows, I know how to setup GUI testing for Linux, I have no idea how to setup GUI testing for both... they are using completely different GUI toolkits.. wouldn't Windows be enough?

**Comment by tomspilman** [comment]
> Wouldn't Windows be enough?

If our goal is for the Pipeline tool to only work reliably on Windows then yes.  Honestly most all MG users use Windows...  maybe 2 or 3% would be on Mac or Linux.  So if it worked right on Windows and totally broken on Linux we would be ok.


> I have no idea how to setup GUI testing for both, they are using completely different GUI toolkits.

Like I said before we have to provide an abstraction that can target either of the GUI toolkits and/or platforms.  Ideally most of it could work via Eto calls which work on both.

We should start with figuring out what sort of tests we want to do then write a testing API that exposes what we need.

**Comment by cra0zy** [comment]
Here is a question I've been trying to figure out an answer to for the past 30 minutes, how do you setup GUI test for something like #5697? If we try too look at it from the code side everything will seem fine. If we try to take screenshots to compare them to, we need to do it for all 3 systems separately, and any change to the default theme would break buildbot screenshots so updating buildbot would become hard and dangerous.

**Comment by tomspilman** [comment]
> If we try to take screenshots to compare them to

I think that is what we would have to do in this case.

> we need to do it for all 3 systems separately, 

Yes... we would need three different source screenshots based on platform.

> any change to the default theme would break buildbot screenshots

True...  but I suspect this would only occur every few months at most.  It should not be a regular occurrence.

When the tests fail a quick review of the captured images would tell us what the issue is.

The bigger problem would be that running the tests locally is hard if you have a different theme on your system.  Maybe we can force a theme under testing for some platforms to avoid that.

**Comment by cra0zy** [comment]
> True... but I suspect this would only occur every few months at most

I don't wanna have to fix tests every few months, plus the more tests are added, the harder it would be fixing them next time the break happens....

> The bigger problem would be that running the tests locally is hard

Running test locally for Linux would be impossible due to the fact that Gtk works differently and has different themes between versions.

**Comment by tomspilman** [comment]
> plus the more tests are added, the harder it would be fixing them next time the break happens

Not really.  We just check the broken tests... and if they are not broken and it is something OS/theme related that changed...  then we either fix that or check in the new test comparison images.  If we do thing right... shouldn't be more than a single folder copy and commit.  Shouldn't take more than a few minutes to do.

Still better than the Pipeline tool being randomly broken and no one noticing for days or weeks.

> Running test locally for Linux would be impossible 

I guess then Linux can only be run on our system or on very specific Linux/GTK versions.

**Comment by cra0zy** [comment]
Instead of having automated tests which would take a long time to write and quite a lot of work to maintain, how about we request that each PR for Pipeline Tool be manually tested on each OS? Yes, it would take longer to merge Pipeline Tool PRs, but it would still require less work than maintaining the tests. Also we could still add PipelineController tests on the side.

**Comment by tomspilman** [comment]
> how about we request that each PR for Pipeline Tool be manually tested on each OS?

I don't think this is reasonable.  Just with the experience we just had with the "Add Existing Item..." changes in https://github.com/MonoGame/MonoGame/pull/6223.  There are at least a dozen different things to test to verify that this feature didn't break things.  I just don't believe a person will do all the tests or even remember what to test.

Back in the day (1997) the testers i worked with had "test scripts".  They would write scripts of all the tests they had to do, step by step, so they would not forget what to test.  Then each day they would break out the nightly build...  and go thru their script testing everything they did the day before.  They did this daily... adding new scripts for new features we added and reporting bugs they found back to us to fix.

This was before there were UI functional testing applications... or at least none that were as cheap as hiring 2 testers.  If we were smarter at the time we would have hired one coder to write code to automate this testing and gotten a huge improvement in quality and reliability of our testing.

Anyway...  i think to get started we could focus on unit tests on internal functions within the Pipeline tool.  Things like `Util.GetRelativePath`, `PipelineProject`, `PipelineProjectParser`, `OutputParser`, etc could be unit tested without any UI.  This could make these bits more robust which can help.

Still i think we have to eventually write a function testing abstraction that can drive the UI.  If i get a weekend free i'll give it a try myself.
